File: src/demo2_apriltag_docking/demo2_apriltag_docking/tag_pose_bridge.py
import math

from apriltag_msgs.msg import AprilTagDetectionArray
from demo2_apriltag_docking.monitor import make_status, shutdown_if_running
from demo2_apriltag_docking.tag_policy import Detection, load_dock_specs, TagGate
from diagnostic_msgs.msg import DiagnosticArray, DiagnosticStatus
from geometry_msgs.msg import PoseStamped
import rclpy
from rclpy.duration import Duration
from rclpy.node import Node
from rclpy.qos import DurabilityPolicy, QoSProfile, ReliabilityPolicy
from rclpy.time import Time
from std_msgs.msg import String
from tf2_ros import Buffer, TransformException, TransformListener
import yaml


def to_policy_detection(message, transform, stamp):
    rotation = transform.transform.rotation
    sin_yaw = 2.0 * (rotation.w * rotation.z + rotation.x * rotation.y)
    cos_yaw = 1.0 - 2.0 * (rotation.y * rotation.y + rotation.z * rotation.z)
    translation = transform.transform.translation
    return Detection(
        tag_id=int(message.id),
        hamming=int(message.hamming),
        decision_margin=float(message.decision_margin),
        stamp=stamp,
        x=float(translation.x),
        y=float(translation.y),
        yaw=math.atan2(sin_yaw, cos_yaw),
    )


def metadata_only(message, stamp):
    return Detection(
        tag_id=int(message.id),
        hamming=int(message.hamming),
        decision_margin=float(message.decision_margin),
        stamp=stamp,
        x=0.0,
        y=0.0,
        yaw=0.0,
    )


def lookup_tag_transform(buffer, target_frame, tag_frame):
    return buffer.lookup_transform(target_frame, tag_frame, Time())


def to_pose_message(transform):
    pose = PoseStamped()
    pose.header = transform.header
    pose.pose.position.x = transform.transform.translation.x
    pose.pose.position.y = transform.transform.translation.y
    pose.pose.position.z = transform.transform.translation.z
    pose.pose.orientation = transform.transform.rotation
    return pose


class TagPoseBridge(Node):

    def __init__(self):
        super().__init__('tag_pose_bridge')
        self._declare_parameters()

        mapping_file = self.get_parameter('dock_mapping_file').value
        if not mapping_file:
            raise ValueError('dock_mapping_file must be set')
        with open(mapping_file, encoding='utf-8') as stream:
            self.specs = load_dock_specs(yaml.safe_load(stream))

        publish_rate = float(self.get_parameter('publish_rate_hz').value)
        self.gate = TagGate(
            specs=self.specs,
            min_margin=float(self.get_parameter('min_decision_margin').value),
            max_hamming=int(self.get_parameter('max_hamming').value),
            confirmations=int(self.get_parameter('confirmations').value),
            confirmation_window=float(
                self.get_parameter('confirmation_window').value
            ),
            publish_period=1.0 / publish_rate,
            loss_timeout=float(self.get_parameter('loss_timeout').value),
            max_translation_jump=float(
                self.get_parameter('max_translation_jump').value
            ),
            max_yaw_jump=math.radians(
                float(self.get_parameter('max_yaw_jump_deg').value)
            ),
        )
        state_qos = QoSProfile(
            depth=1,
            reliability=ReliabilityPolicy.RELIABLE,
            durability=DurabilityPolicy.TRANSIENT_LOCAL,
        )
        self.pose_publisher = self.create_publisher(
            PoseStamped,
            self.get_parameter('output_pose_topic').value,
            10,
        )
        self.state_publisher = self.create_publisher(
            String,
            self.get_parameter('state_topic').value,
            state_qos,
        )
        self.diagnostic_publisher = self.create_publisher(
            DiagnosticArray,
            '/diagnostics',
            10,
        )
        self.tf_buffer = Buffer(cache_time=Duration(seconds=10.0))
        self.tf_listener = TransformListener(self.tf_buffer, self)
        self.subscription = self.create_subscription(
            AprilTagDetectionArray,
            self.get_parameter('detections_topic').value,
            self._on_detections,
            10,
        )
        self.loss_timer = self.create_timer(0.1, self._check_loss)
        self._last_state = None
        self._last_diagnostic_time = None

    def _declare_parameters(self):
        defaults = {
            'dock_mapping_file': '',
            'detections_topic': '/apriltag/detections',
            'output_pose_topic': '/detected_dock_pose',
            'state_topic': '/demo2/tag_state',
            'min_decision_margin': 50.0,
            'max_hamming': 0,
            'confirmations': 3,
            'confirmation_window': 0.5,
            'publish_rate_hz': 10.0,
            'loss_timeout': 0.5,
            'max_translation_jump': 0.25,
            'max_yaw_jump_deg': 20.0,
        }
        for name, value in defaults.items():
            self.declare_parameter(name, value)

    def _on_detections(self, message):
        now = self._now_seconds()
        stamp = message.header.stamp.sec + message.header.stamp.nanosec * 1e-9
        detections = list(message.detections)

        if len(detections) != 1:
            result = self.gate.evaluate(
                [metadata_only(item, stamp) for item in detections],
                now,
            )
            self._report(result.reason, DiagnosticStatus.WARN)
            return

        tag_message = detections[0]
        dock = self.specs.get(int(tag_message.id))
        if dock is None:
            result = self.gate.evaluate([metadata_only(tag_message, stamp)], now)
            self._report(
                result.reason,
                DiagnosticStatus.WARN,
                {'tag_id': tag_message.id},
            )
            return

        try:
            transform = lookup_tag_transform(
                self.tf_buffer,
                message.header.frame_id,
                dock.tag_frame,
            )
        except TransformException as exc:
            self._report(
                'TF_UNAVAILABLE',
                DiagnosticStatus.WARN,
                {'tag_id': tag_message.id, 'error': exc},
            )
            return

        transform_stamp = (
            transform.header.stamp.sec
            + transform.header.stamp.nanosec * 1e-9
        )
        detection = to_policy_detection(tag_message, transform, transform_stamp)
        result = self.gate.evaluate([detection], now)
        if result.accepted:
            self.pose_publisher.publish(to_pose_message(transform))
            self._report(
                'ACCEPTED',
                DiagnosticStatus.OK,
                {
                    'tag_id': detection.tag_id,
                    'dock_id': result.dock.dock_id,
                    'margin': detection.decision_margin,
                },
                refresh_seconds=1.0,
            )
        elif result.reason != 'RATE_LIMITED':
            self._report(
                result.reason,
                DiagnosticStatus.WARN,
                {'tag_id': detection.tag_id},
            )

    def _check_loss(self):
        reason = self.gate.loss_reason(self._now_seconds())
        if reason:
            self._report(reason, DiagnosticStatus.WARN)

    def _report(self, state, level, values=None, refresh_seconds=None):
        now = self._now_seconds()
        repeated = state == self._last_state
        refresh_due = (
            refresh_seconds is not None
            and self._last_diagnostic_time is not None
            and now - self._last_diagnostic_time >= refresh_seconds
        )
        if repeated and not refresh_due:
            return

        if not repeated:
            self.state_publisher.publish(String(data=state))
            self._last_state = state

        diagnostics = DiagnosticArray()
        diagnostics.header.stamp = self.get_clock().now().to_msg()
        diagnostics.status.append(
            make_status(self.get_name(), level, state, values or {})
        )
        self.diagnostic_publisher.publish(diagnostics)
        self._last_diagnostic_time = now

    def _now_seconds(self):
        return self.get_clock().now().nanoseconds * 1e-9


def main(args=None):
    rclpy.init(args=args)
    node = None
    try:
        node = TagPoseBridge()
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    except (OSError, ValueError, yaml.YAMLError) as exc:
        if node is not None:
            node.get_logger().fatal(str(exc))
        else:
            print(f'tag_pose_bridge: {exc}')
    finally:
        if node is not None:
            node.destroy_node()
        shutdown_if_running()


if __name__ == '__main__':
    main()

File: src/demo2_apriltag_docking/src/tag_pose_bridge.cpp
#include "demo2_apriltag_docking/quality_policy.hpp"
#include "demo2_apriltag_docking/tag_policy.hpp"

#include <algorithm>
#include <array>
#include <chrono>
#include <cmath>
#include <cstdio>
#include <stdexcept>
#include <cstdint>
#include <map>
#include <memory>
#include <string>
#include <utility>
#include <vector>

#include <apriltag_msgs/msg/april_tag_detection_array.hpp>
#include <diagnostic_msgs/msg/diagnostic_array.hpp>
#include <diagnostic_msgs/msg/diagnostic_status.hpp>
#include <diagnostic_msgs/msg/key_value.hpp>
#include <geometry_msgs/msg/pose_stamped.hpp>
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/camera_info.hpp>
#include <std_msgs/msg/string.hpp>
#include <tf2/LinearMath/Quaternion.h>
#include <tf2/exceptions.h>
#include <tf2/time.h>
#include <tf2_ros/buffer.h>
#include <tf2_ros/transform_listener.h>

namespace demo2_apriltag_docking
{
namespace
{

double stamp_seconds(const builtin_interfaces::msg::Time & stamp)
{
  return static_cast<double>(stamp.sec) + static_cast<double>(stamp.nanosec) * 1e-9;
}

Detection metadata_only(const apriltag_msgs::msg::AprilTagDetection & message, const double stamp)
{
  return Detection{
    static_cast<int>(message.id), static_cast<int>(message.hamming),
    static_cast<double>(message.decision_margin), stamp, 0.0, 0.0, 0.0};
}

Detection to_policy_detection(
  const apriltag_msgs::msg::AprilTagDetection & message,
  const geometry_msgs::msg::TransformStamped & transform,
  const double stamp)
{
  const auto & rotation = transform.transform.rotation;
  const auto & translation = transform.transform.translation;
  const double sin_yaw = 2.0 * (rotation.w * rotation.z + rotation.x * rotation.y);
  const double cos_yaw = 1.0 - 2.0 * (rotation.y * rotation.y + rotation.z * rotation.z);
  return Detection{
    static_cast<int>(message.id), static_cast<int>(message.hamming),
    static_cast<double>(message.decision_margin), stamp,
    translation.x, translation.y, std::atan2(sin_yaw, cos_yaw)};
}

geometry_msgs::msg::PoseStamped to_pose_message(
  const geometry_msgs::msg::TransformStamped & transform)
{
  geometry_msgs::msg::PoseStamped pose;
  pose.header = transform.header;
  pose.pose.position.x = transform.transform.translation.x;
  pose.pose.position.y = transform.transform.translation.y;
  pose.pose.position.z = transform.transform.translation.z;
  pose.pose.orientation = transform.transform.rotation;
  return pose;
}

}  // namespace

class TagPoseBridge : public rclcpp::Node
{
public:
  TagPoseBridge()
  : Node("tag_pose_bridge"), tf_buffer_(this->get_clock())
  {
    declare_parameter<std::string>("dock_mapping_file", "");
    declare_parameter<std::string>("detections_topic", "/apriltag/detections");
    declare_parameter<std::string>("camera_info_topic", "/camera/camera_info");
    declare_parameter<std::string>("output_pose_topic", "/detected_dock_pose");
    declare_parameter<std::string>("state_topic", "/demo2/tag_state");
    declare_parameter<double>("min_decision_margin", 50.0);
    declare_parameter<int>("max_hamming", 0);
    declare_parameter<int>("confirmations", 3);
    declare_parameter<double>("confirmation_window", 0.5);
    declare_parameter<double>("publish_rate_hz", 10.0);
    declare_parameter<double>("loss_timeout", 0.5);
    declare_parameter<double>("max_translation_jump", 0.25);
    declare_parameter<double>("max_yaw_jump_deg", 20.0);
    declare_parameter<bool>("require_camera_calibration", true);
    declare_parameter<double>("max_detection_age", 0.25);
    declare_parameter<double>("max_tf_age", 0.25);
    declare_parameter<double>("max_future_skew", 0.05);

    const auto mapping_file = get_parameter("dock_mapping_file").as_string();
    if (mapping_file.empty()) {
      throw std::invalid_argument("dock_mapping_file must be set");
    }
    specs_ = load_dock_specs(mapping_file);
    const double publish_rate = get_parameter("publish_rate_hz").as_double();
    if (!std::isfinite(publish_rate) || publish_rate <= 0.0) {
      throw std::invalid_argument("publish_rate_hz must be > 0");
    }
    constexpr double pi = 3.14159265358979323846;
    gate_ = std::make_unique<TagGate>(
      specs_, get_parameter("min_decision_margin").as_double(),
      get_parameter("max_hamming").as_int(), get_parameter("confirmations").as_int(),
      get_parameter("confirmation_window").as_double(), 1.0 / publish_rate,
      get_parameter("loss_timeout").as_double(),
      get_parameter("max_translation_jump").as_double(),
      get_parameter("max_yaw_jump_deg").as_double() * pi / 180.0);

    rclcpp::QoS state_qos(rclcpp::KeepLast(1));
    state_qos.reliable().transient_local();
    pose_publisher_ = create_publisher<geometry_msgs::msg::PoseStamped>(
      get_parameter("output_pose_topic").as_string(), 10);
    state_publisher_ = create_publisher<std_msgs::msg::String>(
      get_parameter("state_topic").as_string(), state_qos);
    diagnostic_publisher_ = create_publisher<diagnostic_msgs::msg::DiagnosticArray>(
      "/diagnostics", 10);

    tf_listener_ = std::make_shared<tf2_ros::TransformListener>(tf_buffer_);
    camera_info_subscription_ = create_subscription<sensor_msgs::msg::CameraInfo>(
      get_parameter("camera_info_topic").as_string(), 10,
      [this](const sensor_msgs::msg::CameraInfo::SharedPtr message) {
        std::array<double, 9> matrix{};
        std::copy(message->k.begin(), message->k.end(), matrix.begin());
        camera_calibration_ = validate_camera_calibration(
          static_cast<int>(message->width), static_cast<int>(message->height),
          matrix, message->distortion_model);
      });
    detection_subscription_ = create_subscription<apriltag_msgs::msg::AprilTagDetectionArray>(
      get_parameter("detections_topic").as_string(), 10,
      [this](const apriltag_msgs::msg::AprilTagDetectionArray::SharedPtr message) {
        on_detections(message);
      });
    loss_timer_ = create_wall_timer(
      std::chrono::milliseconds(100), [this]() {check_loss();});
  }

private:
  double now_seconds() const {return get_clock()->now().seconds();}

  void on_detections(const apriltag_msgs::msg::AprilTagDetectionArray::SharedPtr & message)
  {
    const double now = now_seconds();
    const double stamp = stamp_seconds(message->header.stamp);
    const auto & detections = message->detections;
    if (detections.size() != 1U) {
      std::vector<Detection> metadata;
      metadata.reserve(detections.size());
      for (const auto & item : detections) {
        metadata.push_back(metadata_only(item, stamp));
      }
      const auto result = gate_->evaluate(metadata, now);
      report(to_string(result.reason), diagnostic_msgs::msg::DiagnosticStatus::WARN);
      return;
    }

    const auto & tag_message = detections.front();
    const auto dock_it = specs_.find(static_cast<int>(tag_message.id));
    if (dock_it == specs_.end()) {
      const auto result = gate_->evaluate({metadata_only(tag_message, stamp)}, now);
      report(to_string(result.reason), diagnostic_msgs::msg::DiagnosticStatus::WARN);
      return;
    }

    geometry_msgs::msg::TransformStamped transform;
    try {
      transform = tf_buffer_.lookupTransform(
        message->header.frame_id, dock_it->second.tag_frame, tf2::TimePointZero);
    } catch (const tf2::TransformException & ex) {
      report("TF_UNAVAILABLE", diagnostic_msgs::msg::DiagnosticStatus::WARN);
      RCLCPP_DEBUG(get_logger(), "TF lookup failed: %s", ex.what());
      return;
    }

    const double transform_stamp = stamp_seconds(transform.header.stamp);
    const auto timing = validate_observation_timing(
      now, stamp, transform_stamp, get_parameter("max_detection_age").as_double(),
      get_parameter("max_tf_age").as_double(), get_parameter("max_future_skew").as_double());
    if (!timing.accepted) {
      report(timing.reason, diagnostic_msgs::msg::DiagnosticStatus::WARN);
      return;
    }
    if (get_parameter("require_camera_calibration").as_bool()) {
      if (!camera_calibration_.has_value()) {
        report("CALIBRATION_UNAVAILABLE", diagnostic_msgs::msg::DiagnosticStatus::WARN);
        return;
      }
      if (!camera_calibration_->valid) {
        report(camera_calibration_->reason, diagnostic_msgs::msg::DiagnosticStatus::WARN);
        return;
      }
    }

    const auto result = gate_->evaluate(
      {to_policy_detection(tag_message, transform, transform_stamp)}, now);
    if (result.accepted) {
      pose_publisher_->publish(to_pose_message(transform));
      report("ACCEPTED", diagnostic_msgs::msg::DiagnosticStatus::OK);
    } else if (result.reason != GateReason::kRateLimited) {
      report(to_string(result.reason), diagnostic_msgs::msg::DiagnosticStatus::WARN);
    }
  }

  void check_loss()
  {
    const auto reason = gate_->loss_reason(now_seconds());
    if (reason.has_value()) {
      report(to_string(*reason), diagnostic_msgs::msg::DiagnosticStatus::WARN);
    }
  }

  void report(const std::string & state, const std::uint8_t level)
  {
    const double now = now_seconds();
    const bool repeated = state == last_state_;
    const bool refresh_due = state == "ACCEPTED" &&
      (last_diagnostic_time_ == 0.0 || now - last_diagnostic_time_ >= 1.0);
    if (repeated && !refresh_due) {
      return;
    }
    if (!repeated) {
      std_msgs::msg::String state_message;
      state_message.data = state;
      state_publisher_->publish(state_message);
      last_state_ = state;
    }
    diagnostic_msgs::msg::DiagnosticArray diagnostics;
    diagnostics.header.stamp = get_clock()->now().to_msg();
    diagnostic_msgs::msg::DiagnosticStatus status;
    status.name = get_name();
    status.hardware_id = "demo2_apriltag_docking";
    status.level = level;
    status.message = state;
    diagnostics.status.push_back(status);
    diagnostic_publisher_->publish(diagnostics);
    last_diagnostic_time_ = now;
  }

  std::unordered_map<int, DockSpec> specs_;
  std::unique_ptr<TagGate> gate_;
  tf2_ros::Buffer tf_buffer_;
  std::shared_ptr<tf2_ros::TransformListener> tf_listener_;
  rclcpp::Publisher<geometry_msgs::msg::PoseStamped>::SharedPtr pose_publisher_;
  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr state_publisher_;
  rclcpp::Publisher<diagnostic_msgs::msg::DiagnosticArray>::SharedPtr diagnostic_publisher_;
  rclcpp::Subscription<sensor_msgs::msg::CameraInfo>::SharedPtr camera_info_subscription_;
  rclcpp::Subscription<apriltag_msgs::msg::AprilTagDetectionArray>::SharedPtr detection_subscription_;
  rclcpp::TimerBase::SharedPtr loss_timer_;
  std::optional<CameraCalibration> camera_calibration_;
  std::string last_state_;
  double last_diagnostic_time_{0.0};
};

}  // namespace demo2_apriltag_docking

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  try {
    auto node = std::make_shared<demo2_apriltag_docking::TagPoseBridge>();
    rclcpp::spin(node);
  } catch (const std::exception & ex) {
    fprintf(stderr, "tag_pose_bridge: %s\n", ex.what());
  }
  rclcpp::shutdown();
  return 0;
}
